akka-typed/src/main/scala/akka/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+import akka.japi.function.{ Creator, Function, Predicate }
+import akka.typed.javadsl.Actor.Receive
+import akka.typed.{ Behavior, Signal }
+import ReceiveBuilder._
+import akka.annotation.InternalApi
+
+/**
+ * Used when implementing [[Actor.MutableBehavior]].
+ *
+ * When handling a message or signal, this [[Behavior]] will consider all handlers in the order they were added,
+ * looking for the first handler for which both the type and the (optional) predicate match.
+ *
+ * @tparam T the common superclass of all supported messages.
+ */
+class ReceiveBuilder[T] private (
+  private val messageHandlers: List[Case[T, T]],
+  private val signalHandlers:  List[Case[T, Signal]]
+) {
+
+  def build(): Receive[T] = new BuiltReceive(messageHandlers.reverse, signalHandlers.reverse)
```

Comment:
Given that List (and other Scala collection types) performance is [not very good](http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html) -  there is one extra object for each position, looping has an extra load-cons-from-memory indirection, should we transform it into something else here (thinking `Array`) to not get bad performance for Java Actors? Applies to the other Java PR as well, but didn't think of it until now.